rust/src/lib.rs

```diff
@@ -106,430 +139,387 @@ pub extern "C" fn zcashlc_clear_last_error() {
-            }
-            Err(e) => {
-                return Err(format_err!(
-                    "Error while deriving viewing key from string input: {}",
-                    e
-                ));
-            }
-        };
-        let address = extfvk.default_address().unwrap().1;
-        let address_str = encode_payment_address(network.hrp_sapling_payment_address(), &address);
-        Ok(CString::new(address_str).unwrap().into_raw())
-    });
-    unwrap_exc_or_null(res)
+/// A private utility function to reduce duplication across functions that take an USK
+/// across the FFI. Callers should reproduce the following safety documentation.
+///
+/// # Safety
+///
+/// - `usk_ptr` must be non-null and must point to an array of `usk_len` bytes containing a unified
+///   spending key encoded as returned from the `zcashlc_create_account` or
+///   `zcashlc_derive_spending_key` functions.
+/// - The memory referenced by `usk_ptr` must not be mutated for the duration of the function call.
+/// - The total size `usk_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of pointer::offset.
+unsafe fn decode_usk(
+    usk_ptr: *const u8,
+    usk_len: usize,
+) -> Result<UnifiedSpendingKey, failure::Error> {
+    let usk_bytes = slice::from_raw_parts(usk_ptr, usk_len); //unsafe
+
+    // The remainder of the function is safe.
+    UnifiedSpendingKey::from_bytes(Era::Orchard, usk_bytes).map_err(|e| match e {
+        DecodingError::EraMismatch(era) => format_err!(
+            "Spending key was from era {:?}, but {:?} was expected.",
+            era,
+            Era::Orchard
+        ),
+        e => format_err!(
+            "An error occurred decoding the provided unified spending key: {:?}",
+            e
+        ),
+    })
 }
 
-/// derives a shielded address from the given extended full viewing key.
-/// call zcashlc_string_free with the returned pointer when done using it
+/// Obtains the unified full viewing key for the given binary-encoded unified spending key
+/// and returns the resulting encoded UFVK string.
+///
+/// # Safety
+///
+/// - `usk_ptr` must be non-null and must point to an array of `usk_len` bytes containing a unified
+///   spending key encoded as returned from the `zcashlc_create_account` or
+///   `zcashlc_derive_spending_key` functions.
+/// - The memory referenced by `usk_ptr` must not be mutated for the duration of the function call.
+/// - The total size `usk_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of pointer::offset.
+/// - Call [`zcashlc_string_free`] to free the memory associated with the returned pointer
+///   when you are done using it.
 #[no_mangle]
-pub unsafe extern "C" fn zcashlc_derive_extended_full_viewing_key(
-    extsk: *const c_char,
+pub extern "C" fn zcashlc_spending_key_to_full_viewing_key(
```

Comment:
```suggestion
pub unsafe extern "C" fn zcashlc_spending_key_to_full_viewing_key(
```

Reply:
Deferring this to https://github.com/zcash-hackworks/zcash-light-client-ffi/issues/65#issuecomment-1315988668.